### ISSUE TITLE
Some consensus node only  proposal one miner transaction when there are transactions in the network

### ISSUE THREAD

**Issue by KickSeason** [neo-project/neo#817]
Recently we find that some consensus nodes only proposal one minter transaction.  you can find details on [neoscan](https://neoscan.io) from 3846424  to 3856930. During this time, there only 2 out of 7 consensus nodes proposal non-miner transactions.
And there is an old problem that sometimes one certain consensus node won't proposal transaction except miner transaction. This issue has been discussed before at [#474](https://github.com/neo-project/neo/issues/474) . But now seems not solved.
I will do some test first to find the reason. 
First, I will make certain whether the transactions arrived in the mempool of the problem cn. 
Do you have any idea about this? Is a solution to this currently in progress?
@vncoelho  @igormcoelho @jsolman

**Comment by vncoelho** [comment]
Nice, my friend, good innitiave.

Their mempool will be probably empty, it is a problem in the tx rebroadcasting.
It should be P2P optimizations with extra filter details problems.

**Comment by KickSeason** [comment]
I have tested on testnet.
1. I use a tool keep sending transaction to seed9.ngd.network:20333 via p2p.
2. When I find ngd consensus node produce empty block, I get mempool of the node via rpc  and most 
    of the time get nothing. But sometimes get 1or2 transaction
3. Then I use my tool to send tx to the consensus node directly via p2p. The node works again and relay 
    transactions too and  never produce empty block.
4. Then I  stop sending directly and send transactions to seed9.ngd.network:20333, the issue reproduce.
5 Then I send transaction to seed1 testnet and seed5 testnet, the issue still exists.
6. I restart my consensus node. It works again.

All in all,  I think it seems we have route problem. 
Once a node start, itspeers keep unchanging, right?

**Comment by lock9** [comment]
Good work! 
Now we need to know if the problem is with the node itself, or it's peers. It's funny because it should affect non-CN nodes too, right? 
If it always produces blocks when you send tx to it, it means they are not "processing" these transactions.
It is weird because we know it is receiving network messages 🤔  Maybe it is discarding these transactions?

Did you manage to log the messages received 'per node'? We now need confirmation that:
- TXs messages are not arriving; OR
- TXs are being discarded.

Is there another possible cause? I can only think about these two considering the evidence you provided.

**Comment by vncoelho** [comment]
@KickSeason, it is this. This was the hypotheses, now you tests the fact. Nice work.

In general, the classes `connection`, `peer`, `taskmanager` and `protocolhandle` need a better organization.

The class `connection` has an strange think related to the `ack` state, in which it is set to `false` every time data is sent on the TCP, correct, @erikzhang ?
I believe we need to better arrange the timeouts and this requirement.

**Comment by igormcoelho** [comment]
Interesting observations... perhaps message is received but now internally passed on Akka?

**Comment by KickSeason** [comment]
I will continue testing and record all p2p messages. If the node received transaction message, we can be sure it's akka problem or node internal problem.
@lock9 NGD have tried to reproduce on privatenet but not easy. I will carry on test in testnet.

**Comment by shargon** [comment]
any news @KickSeason ?

**Comment by vncoelho** [comment]
@shargon, can you take a look at me last message above about the `ack` flag?

**Comment by KickSeason** [comment]
no progress. I'm testing on testnet. I can only access one consensus node. I have to wait it reproducing this issue.

**Comment by KickSeason** [comment]
I built a p2p-plugin and filter tx inv and tx message:
[p2ptx.log](https://github.com/neo-project/neo/files/3320158/p2ptx.log)
my code:
    
        public bool OnP2PMessage(Message message)  
        {  
            if (message.Command == "inv") {  
                InvPayload inv = message.GetPayload<InvPayload>();  
                if (inv.Type == InventoryType.TX) {  
                    var hs = inv.Hashes.Where(h => Blockchain.Singleton.GetTransaction(h) == null && !Blockchain.Singleton.MemPool.ContainsKey(h)).ToList();  
                    if (0 < hs.Count())  
                    {  
                        string hashes = "[";
                        foreach (var hash in hs)
                        {
                            if (1 < hashes.Length)
                            {
                                hashes += ",";
                            }
                            hashes += hash.ToString();
                        }
                        hashes += "]";
                        log("inv", $"receive transaction inv, type: tx, count: {hs.Count()}, hashs: {hashes}");
                    }                    
                }
            }
            if (message.Command == "tx") {
                Transaction tx = message.GetTransaction();
                string txid = tx.Hash.ToString();
                log("tx", $"receive transaction {txid}");
            }
            return true;
        }
    
I wonder why there are so many hashes that cn didn't get tx from? I searched those hashes. They aren't on blockchain.
Do you guys have any idea?

**Comment by shargon** [comment]
Those hashes are on know hashes but no received by `inv` or `tx` ?

**Comment by KickSeason** [comment]
`[inv]` With this tag, It is an inv message.
`[tx]` With this, It is a tx message.
I only log the tx inv and tx message.

**Comment by KickSeason** [comment]
Ngd have set up another testnet for repoduce the issue.
I add log into v2.10.2 and deploy on testnet as consensus node to record p2p messages and tasks info. Hope I will help.

**Comment by shargon** [comment]
Could NGD test it too with the last patch https://github.com/neo-project/neo/pull/865?

**Comment by KickSeason** [comment]
We just set up another testnet to try reproducing this issue. But it haven't happened yet.
If we can't reproduce, we can't tell whether it works or not.
We will test this patch a few days later.

**Comment by shargon** [comment]
Thanks for all @KickSeason !

**Comment by superboyiii** [comment]
@shargon It's already happened on Testnet No.2, we're making analysis for consensus log.

**Comment by shargon** [comment]
Good news, the first step is "reproduce it"

**Comment by KickSeason** [comment]
hello guys. I add log in neo based on v2.10.2 and reprodued the issue. Here is the log file:
[2019-07-02.log](https://consensuslog.oss-us-west-1.aliyuncs.com/2019-07-04.log?Expires=1562213865&OSSAccessKeyId=TMP.AgFy059N5WV3qHd0fZUXaJQeYgMQU0H_SKsjHQpRfl939E8mKNKIdLvnRV5FAAAwLAIUE0O5KU-B8oVkBsRPMgx6PIa7JkoCFC2RtASY8nk7NfQSw7rOEpoyambx&Signature=3q1NEYIaTxNI1xwKqNQmw%2FC4qsM%3D)

I took a quick view at the file just now. Now we can be sure the node received tx hash. But it didn't try to get them.
We should do more deep investigate.
If you need more information I will add more log and reproduce

**Comment by shargon** [comment]
> I took a quick view at the file just now. Now we can be sure the node received tx hash. But it didn't try to get them.

It is like a `knownhash` problem?

**Comment by jsolman** [comment]
Hmm, if it receives the transaction but doesn’t think it is valid, but adds it in known hashes, but it later becomes valid, that could be a problem if knownhashes won’t allow it to receive it again. We should check if this is what is happening.

**Comment by jsolman** [comment]
However that would not cause all blocks by that CN to be completely empty of all but the miner transaction. So that may not explain what is happening.

**Comment by KickSeason** [comment]
There are knownhashes count in log file. About 154000 hashes,  it's around 4MB.

**Comment by KickSeason** [comment]
Will add mempool log and test

**Comment by shargon** [comment]
Could you test in parallel, with and without patch?

**Comment by KickSeason** [comment]
You mean the knownhashes fix path?

**Comment by shargon** [comment]
Yes, just for confirm if this solve the problem, at least in the same time frame

**Comment by KickSeason** [comment]
Btw
```
private void OnInvMessageReceived(InvPayload payload)
        {
            UInt256[] hashes = payload.Hashes.Where(p => knownHashes.Add(p)).ToArray();
            if (hashes.Length == 0) return;
            switch (payload.Type)
            {
                case InventoryType.Block:
                    using (Snapshot snapshot = Blockchain.Singleton.GetSnapshot())
                        hashes = hashes.Where(p => !snapshot.ContainsBlock(p)).ToArray();
                        Logger.Info("ProtocolHandler", $"receive inv, type: block, hashes: {hashes.HashesToString()}");
                    break;
                case InventoryType.TX:
                    using (Snapshot snapshot = Blockchain.Singleton.GetSnapshot())
                        hashes = hashes.Where(p => !snapshot.ContainsTransaction(p)).ToArray();
                        Logger.Info("ProtocolHandler", $"receive inv, type: tx, hashes: {hashes.HashesToString()}");
                    break;
            }
            if (hashes.Length == 0) return;
            system.TaskManager.Tell(new TaskManager.NewTasks { Payload = InvPayload.Create(payload.Type, hashes) }, Context.Parent);
        }
````
This is where I  add inv tx log. I think it is nothing to do with knownHases in ProtocolHandler if there is a print log.

**Comment by KickSeason** [comment]
I think we make a great progress. I add more logs and reproduced again in our testnet2.
[2017-07-07.log](https://consensuslog.oss-us-west-1.aliyuncs.com/2019-07-07.log). This all logs of a fault node.
[consensus.log](https://consensuslog.oss-us-west-1.aliyuncs.com/consensus.log). Consensus log I filtered out.
[TaskManagerMailbox.log](https://consensuslog.oss-us-west-1.aliyuncs.com/consensus.log). TaskManagerMailbox log I filtered out.
you can look at the TaskManagerMailbox.log first. It's is stuck by a lot of low priority messages.
![image](https://user-images.githubusercontent.com/22324403/60794126-a878c280-a19b-11e9-9cb1-aa62732a2f5a.png)

**Comment by shargon** [comment]
13.300 low priority messages????

**Comment by KickSeason** [comment]
yes

**Comment by shargon** [comment]
> TaskManagerMailbox.log. TaskManagerMailbox log I filtered out.
Is the same url, this is right?

**Comment by KickSeason** [comment]
Sorry, mistake.

**Comment by shargon** [comment]
I would like to know how many duplicates, we should check the hash before append to the queue

**Comment by shargon** [comment]
How many hours until the issue appear?

**Comment by KickSeason** [comment]
3 days after restart.

**Comment by shargon** [comment]
I would like to know the type of the messages in the queue, something like 
```{ "GetBlock": 1123, "GetPeers": 999 ... }```

is that possible?

Thanks for your work!

**Comment by KickSeason** [comment]
Will do.

**Comment by jsolman** [comment]
We should cache the consensus messages on the CN nodes in case they arrive out of order. Other nodes should cache also for a time and drop duplicates and prevent storms of consensus messages.

**Comment by shargon** [comment]
Yes, we can have a consensus message pool

**Comment by jsolman** [comment]
Also any consensus message from a height less than current block can just be dropped immediately and not forwarded.

**Comment by KickSeason** [comment]
I will lose this because #900 #899 ?
